File: dp05_best_sum.py
def best_sum(target, nums):
    if target == 0:
        return []

    if target < 0:
        return

    shortest = None

    for val in nums:
        rem = target - val
        remRes = best_sum(rem, nums)
        if remRes is not None:
            remRes.append(val)
            if shortest is None or len(shortest) > len(remRes):
                shortest = remRes

    return shortest
        

def best_sum_dy(target, nums, memo={}):
    if target in memo:
        return memo[target]

    if target == 0:
        return []

    if target < 0:
        return

    shortest = None

    for val in nums:
        rem = target - val
        remRes = best_sum_dy(rem, nums, memo)
        if remRes is not None:
            combination = remRes + [val]

            if shortest is None or len(shortest) > len(combination):
                shortest = combination

    memo[target] = shortest
    return shortest

def main():
    # print(best_sum(7, [5, 3, 4, 7])) 
    # print(best_sum(8, [2, 3, 5])) 
    # print(best_sum(8, [1, 4, 5])) 
    # print(best_sum(100, [1, 2, 5, 25])) 

    print(best_sum_dy(7, [5, 3, 4, 7], {})) 
    print(best_sum_dy(8, [2, 3, 5], {})) 
    print(best_sum_dy(8, [1, 4, 5], {})) 
    print(best_sum_dy(9, [5], {})) 
    print(best_sum_dy(100, [1, 2, 5, 25], {})) 


main()

File: dp05_bestSum.go
package main

import "fmt"

func main() {
	// fmt.Println(bestSum(7, []int{5, 4, 7}))
	// fmt.Println(bestSum(8, []int{2, 3, 5}))
	// fmt.Println(bestSum(8, []int{1, 4, 5}))
	// fmt.Println(bestSum(8, []int{5}))
	// fmt.Println(bestSum(100, []int{1, 2, 5, 25}))

	fmt.Println(bestSumDy(7, []int{5, 4, 7}, map[int][]int{}))
	fmt.Println(bestSumDy(8, []int{2, 3, 5}, map[int][]int{}))
	fmt.Println(bestSumDy(8, []int{1, 4, 5}, map[int][]int{}))
	fmt.Println(bestSumDy(8, []int{5}, map[int][]int{}))
	fmt.Println(bestSumDy(100, []int{1, 2, 5, 25}, map[int][]int{}))
}

func bestSum(target int, nums []int) []int {
	if target == 0 {
		return []int{}
	}

	if target < 0 {
		return nil
	}

	// shortest := new([]int)
	var shortest []int

	for _, val := range nums {
		rem := target - val

		remRes := bestSum(rem, nums)
		if remRes != nil {
			cobination := append(remRes, val)
			if shortest == nil || len(cobination) < len(shortest) {
				shortest = cobination
			}
		}
	}

	return shortest
}

func bestSumDy(target int, nums []int, memo map[int][]int) []int {
	if val, ok := memo[target]; ok {
		return val
	}

	if target == 0 {
		return []int{}
	}

	if target < 0 {
		return nil
	}

	// shortest := new([]int)
	var shortest []int

	for _, val := range nums {
		rem := target - val

		remRes := bestSumDy(rem, nums, memo)
		if remRes != nil {
			cobination := append(remRes, val)
			if shortest == nil || len(cobination) < len(shortest) {
				shortest = cobination
			}
		}
	}

	memo[target] = shortest
	return shortest
}
